### ISSUE TITLE
Crafting is slow

### ISSUE THREAD

**Issue by demipixel** [andrewrk/mineflayer#257]
I've had experience and @roblabla suggested that I make a post about since he believes we're doing a lot of unnecessary work.

In some cases I can craft faster than the bot.

**Comment by roblabla** [comment]
I believe we shouldn't be waiting after each click for the transaction to complete. Instead, we could send all clicks at once (except the final one), check if we got any errors, and if we didn't send the final click.

**Comment by rom1504** [comment]
Yeah I guess we could optimize it. It might even not be very hard since nobody tried yet afaik (craft was broken not long ago).

**Comment by rom1504** [comment]
related to #129

**Comment by rom1504** [comment]
244fc7b76eb2a08d54411b5bd6e02d10cdb4aa61 improves bot.transfer by left clicking when possible instead of always right clicking, bot.craft could probably be improved in a similar way (using different kind of clicks)

**Comment by FeldrinH** [comment]
In minecraft 1.12 and above, fast crafting could be achieved by using [craft recipe request packet](https://wiki.vg/Protocol#Craft_Recipe_Request) to move items into the crafting grid (multiple consecutive packets can be sent to increase the number of items in the crafting grid), and a shift click to withdraw all items at once.

**Comment by sefirosweb** [comment]
Checking this issue, the slow performance comes for waiting to open window,
I measured all steps to craft items and this step are the 98% of lost time

https://github.com/PrismarineJS/mineflayer/blob/20862737c733fecee92b78fa12ea360003126290/lib/plugins/craft.js#L24

What I don't quite understand is that the FIRST execution takes only 22ms,

But if another crafting is requested shortly (example x 64 items) the second request takes 900ms~ why? I think it is due to a windowOpen failure

I attach the perform test:

![image](https://user-images.githubusercontent.com/20754836/215291448-e38dce4d-3966-4f9d-b7f6-56f8fc51a4a6.png)

Of course it could be optimized to not close the window and finish crafting, but I think it would be important to see why it takes so long to open the window

**Comment by rom1504** [comment]
there are at least 2 PRs solving this issue but they didn't get finished

you could pick one and finish it

**Comment by sefirosweb** [comment]
@rom1504 yep i'm on it, only are notifying on people if know why happens to save time =P

**Comment by sefirosweb** [comment]
fixed bug! now flies!
